File: fs/src/test/java/l/files/fs/PathHierarchyTest.java
package l.files.fs;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import java.nio.file.Paths;
import java.util.Collection;
import java.util.List;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static java.util.stream.Collectors.toList;
import static l.files.fs.PathKt.hierarchy;
import static org.junit.Assert.assertEquals;

@RunWith(Parameterized.class)
public final class PathHierarchyTest {

    private final java.nio.file.Path path;
    private final List<String> expectedHierarchy;

    public PathHierarchyTest(String path, List<String> expectedHierarchy) {
        this.path = Paths.get(path);
        this.expectedHierarchy = expectedHierarchy;
    }

    @Parameters(name = "\"{0}\".hierarchy() == \"{1}\"")
    public static Collection<Object[]> paths() {
        return asList(new Object[][]{
            {"", singletonList("")},
            {"/", singletonList("/")},
            {"/a/b/c", asList("/a/b/c", "/a/b", "/a", "/")},
        });
    }

    @Test
    public void test() {
        assertEquals(
            expectedHierarchy.stream().map(Paths::get).collect(toList()),
            hierarchy(path)
        );
    }
}

File: fs/src/test/kotlin/l/files/fs/PathHierarchyTest.kt
package l.files.fs

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import java.nio.file.Paths

@RunWith(Parameterized::class)
class PathHierarchyTest(
  path: String,
  expectedHierarchy: List<String>
) {

  private val path = Paths.get(path)
  private val expectedHierarchy = expectedHierarchy.map(Paths::get)

  @Test
  fun test() {
    assertEquals(expectedHierarchy, path.hierarchy())
  }

  companion object {
    @JvmStatic
    @Parameters(name = "\"{0}\".hierarchy() == \"{1}\"")
    fun paths() = listOf(
      arrayOf("", listOf("")),
      arrayOf("/", listOf("/")),
      arrayOf("/a/b/c", listOf("/a/b/c", "/a/b", "/a", "/"))
    )
  }
}
